### PR TITLE
TST: Add list of tuples pd.factorize test

### LINKED ISSUE #9454: factorize fails for list of tuples
It's not clear from the documentation for factorize what datatype is expected for the values. But I assume that any list of hashables should work (specifically, a list of tuples). 

Factorize indeed works for a list of tuples as long as the lens of all the tuples are not identical, but fails the moment all tuples have the same length. (Looks like there is some inference about the structure of the values that shouldn't be happening.)

``` python
import pandas as pd
pd.factorize([(1, 1), (1, 2), (0, 0), (1, 2), 'nonsense']) # This works

(array([0, 1, 2, 1, 3]), array([(1, 1), (1, 2), (0, 0), 'nonsense'], dtype=object))

pd.factorize([(1, 1), (1, 2), (0, 0), (1, 2), (1, 2, 3)]) # This also works.

pd.factorize([(1, 1), (1, 2), (0, 0), (1, 2)]) # <-- fails
```

```
ValueError                                Traceback (most recent call last)
<ipython-input-22-3ca8ec02e16c> in <module>()
      1 print pd.factorize([(1, 1), (1, 2), (0, 0), (1, 2), 'nonsense'])
----> 2 print pd.factorize([(1, 1), (1, 2), (0, 0), (1, 2)])

/usr/local/lib/python2.7/dist-packages/pandas/core/algorithms.pyc in factorize(values, sort, order, na_sentinel)
    132     table = hash_klass(len(vals))
    133     uniques = vec_klass()
--> 134     labels = table.get_labels(vals, uniques, 0, na_sentinel)
    135 
    136     labels = com._ensure_platform_int(labels)

/usr/local/lib/python2.7/dist-packages/pandas/hashtable.so in pandas.hashtable.Int64HashTable.get_labels (pandas/hashtable.c:8575)()

ValueError: Buffer has wrong number of dimensions (expected 1, got 2)
```

pandas 0.15.2

### PR DESCRIPTION
**Pull request by Licht-T**
- [x] closes #9454
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/test_algos.py [hunk 1]
```diff
@@ -190,6 +190,34 @@ def test_factorize_nan(self):
...
+    ])
+    def test_factorize_tuple_list(self, data, expected_label, expected_level):
+        # GH9454
+        result = pd.factorize(data)
+
+        tm.assert_numpy_array_equal(result[0],
+                                    np.array(expected_label))
+
+        expected_level_array = np.empty(len(expected_level), dtype=object)
+        expected_level_array[:] = expected_level
```
Comment (jschendel) [inline pandas/tests/test_algos.py]: You can use `pandas.core.common._asarray_tuplesafe` to get a numpy array of tuples, i.e. `expected_level =  _asarray_tuplesafe(expected_level)`

**Comment by Licht-T** [comment]
@jschendel Thanks! Fixed!

File: pandas/tests/test_algos.py [hunk 2]
```diff
@@ -190,6 +191,33 @@ def test_factorize_nan(self):
...
+            [0, 1, 2, 1],
+            [(1, 1), (1, 2), (0, 0)]
+        )
+    ])
+    def test_factorize_tuple_list(self, data, expected_label, expected_level):
+        # GH9454
+        result = pd.factorize(data)
+
+        tm.assert_numpy_array_equal(result[0],
+                                    np.array(expected_label, dtype=np.int_))
```
Comment (jschendel) [inline pandas/tests/test_algos.py]: I think `dtype=np.intp` should fix the failure on appveyor.

Comment (Licht-T) [reply pandas/tests/test_algos.py]: Ah...! Thanks @jschendel!

**Comment by jreback** [comment]
thanks @Licht-T

**Pull request merged**